### ISSUE TITLE
Confirm to access registration

### ISSUE THREAD

**Issue by nikcorg** [volatiletech/authboss#191]
TL;DR Is it possible to create a confirm-to-register type of new user flow using Authboss?

To elaborate slightly, what we're looking to do is:

1. Collect email address
2. Generate single-use confirmation token to deliver to email collected in step 1.
3. Collect more information post-confirmation
4. Create new user account

I'm new to both Golang and Authboss and couldn't immediately figure out how to accomplish this without implementing it ourselves. If we indeed need to implement this ourselves, as Authboss already has facilities for generating Confirmation tokens and sending email, can we hook into those somehow?

**Comment by aarondl** [comment]
This is possible by stealing most of confirms code. You can reuse its interface (ConfirmableUser) and it's generate token function. But you would have to create a new module for the flow that you want I think. Most of what would be shared is exported AFAIK.

**Comment by nikcorg** [comment]
@aarondl thanks. I'll have a stab at it and if something (elegant and working) comes out of it, I'll definitely consider contributing back :-)

**Comment by nikcorg** [comment]
@aarondl I managed to achieve what we wanted, but the implementation is not ideal.

In the end I went with

1. Copy most of `confirm` into a new module
2. Fork `register` so that I can uncomment the router bindings
3. Bind the routes for `register` in the new `confirm2register` module, augmenting the `Get` and `Post` handlers in register

This led me to thinking that the autosetup is perhaps not ideal and the approach used in the 2FA modules, requiring an explicit call to `module.Setup()` enables greater flexibility.

I get that Authboss is supposed to be very plug&play-ish, but I think there are good arguments for not having side-effects from simply importing a package, such as the call to `authboss.RegisterModule()` in the `init` in most Authboss modules.

For instance I think I wouldn't have had to duplicate nearly as much code as I now did, if I could've simply composed my new `confirm2register` by augmenting the handlers in `confirm` and `register`.

I'm rambling a bit, but there's some food for thought in there. I could definitely pull together a PR better showcasing what I mean, if there is interest.

**Comment by aarondl** [comment]
I absolutely agree that the side effect imports are a mistake hence the new modules not doing that. Unfortunately it's now a breaking change to change it :( Probably should have done it in v2, now we've got to wait until v3.

**Comment by nikcorg** [comment]
Good to hear that you agree. Is there a roadmap for Authboss or is it simply as needs arise or community contributions?

I could definitely be interested in working on this, but wouldn't want to waste work if there is already a plan and work in progress happening elsewhere.

**Comment by aarondl** [comment]
There's no roadmap at the moment. I consider authboss "complete" for what it's supposed to offer. The dev branch has a few things I'll eventually push out is all.

Revving major versions and introducing breaking changes like that frequently is a bit more problematic than in other languages because `go get` so I like to move a bit slower on that front. But I do encourage vendoring etc. :\ Tough call.

**Comment by nikcorg** [comment]
I was going to come here with my great innovation of providing a flag to `authboss.Init` which prevents automatic invoking of `module.Init`, but then I read the source code and noticed `authboss.Init` can already take a list of modules to initialise, which is perfectly sufficient. 

Thanks for chatting and the help, and thanks for Authboss!
